### PR TITLE
Don't deserialize Conway outputs with pointer addresses

### LINKED ISSUE #3161: prevent ptr address from being created in conway
Starting in the conway ledger era, do not allow the creation of transaction outputs that contain pointer addresses. Implementation options include: 1) adding a new predicate failure, perhaps the `UTXO` rule, 2) use the version CBOR to prevent pointer address from being deserialized starting in major protocol version 9.

### PR DESCRIPTION
**Pull request by teodanciu**
Closes https://github.com/input-output-hk/cardano-ledger/issues/3161

### CODE REVIEW

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/TxOut.hs [hunk 1]
```diff
@@ -435,14 +432,28 @@ instance
...
+  ( EraTxOut era,
+    Core.TxOut era ~ BabbageTxOut era,
+    FromCBOR (Annotator (Script era))
   ) =>
   FromCBOR (BabbageTxOut era)
   where
-  fromCBOR = fromCborTxOutWithAddr fromCborBackwardsBothAddr
+  fromCBOR = do
+    txOut <- fromCborTxOutWithAddr fromCborBackwardsBothAddr
+    whenDecoderVersionAtLeast (natVersion @9) $ do
```
Comment (JaredCorduan) [inline eras/babbage/impl/src/Cardano/Ledger/Babbage/TxOut.hs:445]: thinking out loud here, maybe we could leave a comment which points to the ADR I just made? :thinking: 

https://github.com/input-output-hk/cardano-ledger/pull/3192

**Review by JaredCorduan — APPROVED** [review]
The change looks great, thank you! We should update the changelog, and also maybe leave a bit more of a trail about why this change was needed (below I suggested a comment linking to the ADR that I just put up, which in turn links to the CIP...).

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/TxOut.hs [hunk 2]
```diff
@@ -435,14 +432,28 @@ instance
...
+    pure txOut
+
+checkNotPtrAddr ::
+  forall era s.
+  (EraTxOut era) =>
+  Core.TxOut era ->
+  Decoder s ()
+checkNotPtrAddr txOut = do
+  let addr = txOut ^. addrTxOutL
+  if (isPtr addr)
```
Comment (lehins) [inline eras/babbage/impl/src/Cardano/Ledger/Babbage/TxOut.hs:454]: Redundant parenthesis
```suggestion
  if isPtr addr
```

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/TxOut.hs [hunk 3]
```diff
@@ -435,14 +432,28 @@ instance
...
+    Core.TxOut era ~ BabbageTxOut era,
+    FromCBOR (Annotator (Script era))
   ) =>
   FromCBOR (BabbageTxOut era)
   where
-  fromCBOR = fromCborTxOutWithAddr fromCborBackwardsBothAddr
+  fromCBOR = do
+    txOut <- fromCborTxOutWithAddr fromCborBackwardsBothAddr
+    whenDecoderVersionAtLeast (natVersion @9) $ do
+      checkNotPtrAddr txOut
```
Comment (lehins) [inline eras/babbage/impl/src/Cardano/Ledger/Babbage/TxOut.hs]: There is no need to create helper function for everything. It is better to write it like so:
```suggestion
      when (isPtr (txOut ^. addrTxOutL)) $
        fail "Pointer addresses are not supported starting with protocol version 9"
```

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/TxOut.hs [hunk 4]
```diff
@@ -435,14 +432,28 @@ instance
...
+checkNotPtrAddr ::
+  forall era s.
+  (EraTxOut era) =>
+  Core.TxOut era ->
+  Decoder s ()
+checkNotPtrAddr txOut = do
+  let addr = txOut ^. addrTxOutL
+  if (isPtr addr)
+    then fail "Pointer addresses not supported starting with version 9"
+    else pure ()
```
Comment (lehins) [inline eras/babbage/impl/src/Cardano/Ledger/Babbage/TxOut.hs]: Not needed
```suggestion
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Era.hs [hunk 5]
```diff
@@ -25,12 +25,12 @@ import Cardano.Ledger.Shelley.Rules
...
 -- =====================================================
 
--- | The Babbage era
+-- | The Conway era
 data ConwayEra c
 
 instance CC.Crypto c => Era (ConwayEra c) where
   type EraCrypto (ConwayEra c) = c
-  type ProtVerLow (ConwayEra c) = 8
+  type ProtVerLow (ConwayEra c) = 9
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Era.hs:33]: Could you also update the upper version in babbage: https://github.com/input-output-hk/cardano-ledger/blob/6b41318acd067a38f3f12faa2397a9f81921ded2/eras/babbage/impl/src/Cardano/Ledger/Babbage/Era.hs#L36-L37

```diff
+ type ProtVerHigh (BabbageEra c) = 8
```

File: eras/conway/test-suite/src/Test/Cardano/Ledger/Conway/Serialisation/Roundtrip.hs [hunk 6]
```diff
@@ -23,5 +32,25 @@ allprops =
...
     ]
+
+ptrAddrTxOutNotDeserialized :: Property
+ptrAddrTxOutNotDeserialized =
+  forAll (genPtrAddress @(EraCrypto Conway)) $ \ptrAddr txOut ->
+    do
+      let ptrAddrTxOut = txOut & addrTxOutL .~ ptrAddr
+      let messageMatch s = "Pointer addresses not supported starting with version 9" `isInfixOf` s
+      property $
+        case (roundTrip (natVersion @9) (cborTrip @(BabbageTxOut Conway)) ptrAddrTxOut) of
```
Comment (lehins) [inline eras/conway/test-suite/src/Test/Cardano/Ledger/Conway/Serialisation/Roundtrip.hs]: Redundant parenthesis
```suggestion
        case roundTrip (natVersion @9) (cborTrip @(BabbageTxOut era)) ptrAddrTxOut of
```

File: eras/conway/test-suite/src/Test/Cardano/Ledger/Conway/Serialisation/Roundtrip.hs [hunk 7]
```diff
@@ -23,5 +32,25 @@ allprops =
...
     [ testProperty "ConwayGenesis" $
-        roundTripCborExpectation @(ConwayGenesis (EraCrypto e)) (eraProtVerHigh @e)
+        roundTripCborExpectation @(ConwayGenesis (EraCrypto e)) (eraProtVerHigh @e),
+      testProperty "conway/TxOut with pointer addresses" ptrAddrTxOutNotDeserialized,
+      testProperty "conway/TxOut without pointer addresses" nonPtrAddrTxOutDeserialized
     ]
+
+ptrAddrTxOutNotDeserialized :: Property
+ptrAddrTxOutNotDeserialized =
+  forAll (genPtrAddress @(EraCrypto Conway)) $ \ptrAddr txOut ->
```
Comment (lehins) [inline eras/conway/test-suite/src/Test/Cardano/Ledger/Conway/Serialisation/Roundtrip.hs]: In order to make this test forward compatible we need to expose the `era` parameter and not restrict it to `Conway` only.  After all this test should pass for all eras that use `BabbageTxOut` .

```suggestion
ptrAddrTxOutNotDeserialized :: EraTxOut era => BabbageTxOut era -> Property
ptrAddrTxOutNotDeserialized txOut =
  forAll (genPtrAddress @era) $ \ptrAddr ->
```

File: eras/conway/test-suite/src/Test/Cardano/Ledger/Conway/Serialisation/Roundtrip.hs [hunk 8]
```diff
@@ -23,5 +32,25 @@ allprops =
...
+      let ptrAddrTxOut = txOut & addrTxOutL .~ ptrAddr
+      let messageMatch s = "Pointer addresses not supported starting with version 9" `isInfixOf` s
+      property $
+        case (roundTrip (natVersion @9) (cborTrip @(BabbageTxOut Conway)) ptrAddrTxOut) of
+          Left (RoundTripFailure _ _ _ _ _ _ (Just e)) -> messageMatch (show e)
+          _ -> False
+
+nonPtrAddrTxOutDeserialized :: Property
+nonPtrAddrTxOutDeserialized =
+  forAll (genNonPtrAddress @(EraCrypto Conway)) $ \nonPtrAddr txOut ->
```
Comment (lehins) [inline eras/conway/test-suite/src/Test/Cardano/Ledger/Conway/Serialisation/Roundtrip.hs]: Same comments as for the above property test applies here.

File: eras/conway/test-suite/src/Test/Cardano/Ledger/Conway/Serialisation/Roundtrip.hs [hunk 9]
```diff
@@ -23,5 +32,25 @@ allprops =
...
+ptrAddrTxOutNotDeserialized :: Property
+ptrAddrTxOutNotDeserialized =
+  forAll (genPtrAddress @(EraCrypto Conway)) $ \ptrAddr txOut ->
+    do
+      let ptrAddrTxOut = txOut & addrTxOutL .~ ptrAddr
+      let messageMatch s = "Pointer addresses not supported starting with version 9" `isInfixOf` s
+      property $
+        case (roundTrip (natVersion @9) (cborTrip @(BabbageTxOut Conway)) ptrAddrTxOut) of
+          Left (RoundTripFailure _ _ _ _ _ _ (Just e)) -> messageMatch (show e)
+          _ -> False
```
Comment (lehins) [inline eras/conway/test-suite/src/Test/Cardano/Ledger/Conway/Serialisation/Roundtrip.hs]:
```suggestion
          _ -> counterexample "This was not suppose to parse" False
```

File: eras/conway/test-suite/src/Test/Cardano/Ledger/Conway/Serialisation/Roundtrip.hs [hunk 10]
```diff
@@ -23,5 +32,25 @@ allprops =
...
+      testProperty "conway/TxOut with pointer addresses" ptrAddrTxOutNotDeserialized,
+      testProperty "conway/TxOut without pointer addresses" nonPtrAddrTxOutDeserialized
     ]
+
+ptrAddrTxOutNotDeserialized :: Property
+ptrAddrTxOutNotDeserialized =
+  forAll (genPtrAddress @(EraCrypto Conway)) $ \ptrAddr txOut ->
+    do
+      let ptrAddrTxOut = txOut & addrTxOutL .~ ptrAddr
+      let messageMatch s = "Pointer addresses not supported starting with version 9" `isInfixOf` s
```
Comment (lehins) [inline eras/conway/test-suite/src/Test/Cardano/Ledger/Conway/Serialisation/Roundtrip.hs]: If we don't add these explanation we'll not know why exactly our test is failing
```suggestion
      let messageMatch s = counterexample ("Unexpected decoder failure message: " ++ s) $
                                              "Pointer addresses not supported starting with version 9" `isInfixOf` s
```

File: eras/conway/test-suite/src/Test/Cardano/Ledger/Conway/Serialisation/Roundtrip.hs [hunk 11]
```diff
@@ -23,5 +32,25 @@ allprops =
...
+      property $
```
Comment (lehins) [inline eras/conway/test-suite/src/Test/Cardano/Ledger/Conway/Serialisation/Roundtrip.hs]: No longer needed with my other comments
```suggestion
```

File: eras/conway/test-suite/src/Test/Cardano/Ledger/Conway/Serialisation/Roundtrip.hs [hunk 12]
```diff
@@ -23,5 +32,25 @@ allprops =
...
+        case (roundTrip (natVersion @9) (cborTrip @(BabbageTxOut Conway)) ptrAddrTxOut) of
+          Left (RoundTripFailure _ _ _ _ _ _ (Just e)) -> messageMatch (show e)
+          _ -> False
+
+nonPtrAddrTxOutDeserialized :: Property
+nonPtrAddrTxOutDeserialized =
+  forAll (genNonPtrAddress @(EraCrypto Conway)) $ \nonPtrAddr txOut ->
+    do
+      let nonPtrAddrTxOut = txOut & addrTxOutL .~ nonPtrAddr
+      property $ isRight $ (roundTrip (natVersion @9) (cborTrip @(BabbageTxOut Conway)) nonPtrAddrTxOut)
```
Comment (lehins) [inline eras/conway/test-suite/src/Test/Cardano/Ledger/Conway/Serialisation/Roundtrip.hs]: Redundant parenthesis, also isRight will not tell us anything why this test fails, if it does:
```suggestion
      counterexample ("Unexpected decoder failure: " ++ show err) $
      case roundTrip (natVersion @9) (cborTrip @(BabbageTxOut era)) nonPtrAddrTxOut of
        Left err ->  False
        Right  -> True
```

File: libs/cardano-ledger-core/src/Cardano/Ledger/Address.hs [hunk 13]
```diff
@@ -401,6 +379,10 @@ isBootstrapRedeemer :: BootstrapAddress c -> Bool
 isBootstrapRedeemer (BootstrapAddress (Byron.Address _ _ Byron.ATRedeem)) = True
 isBootstrapRedeemer _ = False
 
+isPtr :: Addr era -> Bool
+isPtr (Addr _ _ (StakeRefPtr _)) = True
+isPtr _ = False
```
Comment (lehins) [inline libs/cardano-ledger-core/src/Cardano/Ledger/Address.hs]: Consider a little more descriptive name eg. `isStakePtrAddr`?

**Review by lehins — APPROVED** [review]
This is great! Few recommendations, otherwise I see no blockers.

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/TxOut.hs (hunk 3 above)
Comment (teodanciu) [reply eras/babbage/impl/src/Cardano/Ledger/Babbage/TxOut.hs]: Yep, I think the scala-tendencies took over!  Your version is way easier to read, thanks

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/TxOut.hs (hunk 1 above)
Comment (teodanciu) [reply eras/babbage/impl/src/Cardano/Ledger/Babbage/TxOut.hs:445]: Yep, nice idea!  Will update

File: libs/cardano-ledger-core/src/Cardano/Ledger/Address.hs (hunk 13 above)
Comment (JaredCorduan) [reply libs/cardano-ledger-core/src/Cardano/Ledger/Address.hs]: elsewhere, these are described as "pointer addresses", so I would probably go with `isPtrAddr`. there isn't a notion of "payment pointer address", for example.

File: eras/conway/test-suite/src/Test/Cardano/Ledger/Conway/Serialisation/Roundtrip.hs (hunk 7 above)
Comment (teodanciu) [reply eras/conway/test-suite/src/Test/Cardano/Ledger/Conway/Serialisation/Roundtrip.hs]: Hey @lehins , I'm only able to pass the era as a parameter with these signature: 
```
ptrAddrTxOutNotDeserialized ::
  forall e.
  ( EraTxOut e,
    TxOut e ~ BabbageTxOut e,
    Mock (EraCrypto e),
    Arbitrary (Value e),
    Arbitrary (Script e),
    Show (Script e)
  ) =>
  BabbageTxOut e ->
  Property
```

which then I have to repeat in the top-level function .
Is there a better way? 

Also: these functions get invoked ultimately from  conway `Main`, like this: 
```
Roundtrip.allprops @Conway
```

So passing the era as a parameter will come in handy if in the future eras we will explicitly call this function passing the new era as a parameter  - that's what you meant, correct?

Comment (lehins) [reply eras/conway/test-suite/src/Test/Cardano/Ledger/Conway/Serialisation/Roundtrip.hs]: That's right, that's exactly what I meant. Also you can simplify this a bit:
```haskell
ptrAddrTxOutNotDeserialized ::
  forall e.
  ( EraTxOut e,
    Mock (EraCrypto e),
    Arbitrary (Value e),
    Arbitrary (Script e),
    Show (Script e)
  ) =>
  TxOut e ->
  Property
```

**Comment by JaredCorduan** [comment]
@teodanciu we should mention this in the changelog

**Pull request merged**